overlord/snapstate/refresh.go

```diff
@@ -0,0 +1,213 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package snapstate
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/snapcore/snapd/cmd/snaplock"
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/snap"
+)
+
+func genericRefreshCheck(info *snap.Info, canAppRunDuringRefresh func(app *snap.AppInfo) bool) error {
+	// Grab per-snap lock to prevent new processes from starting. This is
+	// sufficient to perform the check, even though individual processes
+	// may fork or exit, we will have per-security-tag information about
+	// what is running.
+	lock, err := snaplock.OpenLock(info.SnapName())
+	if err != nil {
+		return err
+	}
+	defer lock.Close()
+	if err := lock.Lock(); err != nil {
+		return err
+	}
+
+	var busyAppNames []string
+	var busyHookNames []string
+	var busyPIDs []int
+
+	canHookRunDuringRefresh := func(hook *snap.HookInfo) bool {
+		return false
+	}
+
+	for name, app := range info.Apps {
+		if canAppRunDuringRefresh(app) {
+			continue
+		}
+		PIDs, err := pidsOfSecurityTag(app.SecurityTag())
+		if err != nil {
+			return err
+		}
+		if len(PIDs) > 0 {
+			busyAppNames = append(busyAppNames, name)
+			busyPIDs = append(busyPIDs, PIDs...)
+		}
+	}
+
+	for name, hook := range info.Hooks {
+		if canHookRunDuringRefresh(hook) {
```

Comment:
Thinking about this - do we know already when hooks can run during refreshes? If we don't have any use-case yet, maybe this is a case of YAGNI? I mean, we can always add it if we need it.

Reply:
Currently hooks cannot run during refreshes. This form is just for consistency with app checks but we must check that no hooks are running.